### PR TITLE
Add root chain transaction hash to InvalidExit and UnchallengedExit events

### PR DESCRIPTION
**Pull request by okalouti**
#1480 

## Overview

This adds a `rootchain_txhash` property to the `invalid_exit` and `unchallenged_exit` events, referring to the Ethereum transaction hash of the exit in question.

As seen by running the cabbage tests locally: 

```json
{
        "details": {
          "amount": 12000000000000000000,
          "currency": "0x0000000000000000000000000000000000000000",
          "eth_height": 899,
          "name": "unchallenged_exit",
          "owner": "0x37d5c1fbfa32b6582b425d5ea8d724478dac0aee",
          "root_chain_txhash": "0x5ebfbd6caff8d53e74590a646ffa5018e8776445c78ed735e29bdb52e55edd9c",
          "utxo_pos": 2001000000000
        },
        "event": "unchallenged_exit"
      },
      {
        "details": {
          "amount": 12000000000000000000,
          "currency": "0x0000000000000000000000000000000000000000",
          "eth_height": 899,
          "name": "invalid_exit",
          "owner": "0x37d5c1fbfa32b6582b425d5ea8d724478dac0aee",
          "root_chain_txhash": "0x5ebfbd6caff8d53e74590a646ffa5018e8776445c78ed735e29bdb52e55edd9c",
          "utxo_pos": 2001000000000
        },
        "event": "invalid_exit"
      }

```

## Changes

- Changes in `exit_processor/exit_info.ex` to persist the root chain transaction hash in the database. 
- Changes in `omg_watcher/event.ex` to add root chain transaction hash to the event struct.

## Testing

- No tests added but exit test helper updated to include a transaction hash of `<<0::256`. Open question whether we'd like a more advanced logic there in this PR. 

Note to @dmitrydao to update documentation [here](https://docs.omg.network/byzantine-conditions) and [here](https://docs.omg.network/challenging-exits) once this is merged.

### CODE REVIEW

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex [hunk 1]
```diff
@@ -37,10 +46,19 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
...
           is_active: boolean(),
-          eth_height: pos_integer()
+          eth_height: pos_integer(),
+          root_chain_txhash: Crypto.hash_t()
         }
 
-  def new(contract_status, %{eth_height: eth_height, call_data: %{output_tx: txbytes}, exit_id: exit_id} = event) do
+  def new(
+        contract_status,
+        %{
```
Comment (InoMurko) [inline apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:55]: maybe now is a good time to re-approach this pattern matching.

event is a module defined struct, you can access it without pattern matching it in the function header - this way we can reduce the module bloat that was accumulated

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex [hunk 2]
```diff
@@ -80,20 +99,23 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
            exit_id: exit_id,
            exiting_txbytes: exiting_txbytes,
            is_active: is_active,
-           eth_height: eth_height
+           eth_height: eth_height,
+           root_chain_txhash: root_chain_txhash
```
Comment (InoMurko) [inline apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:103]: same as above.
`event. root_chain_txhash`
you can also remove these guards, I don't know who are they protecting us from.

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex [hunk 3]
```diff
@@ -108,12 +130,13 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
...
```
Comment (InoMurko) [inline apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:134]: same as above

File: apps/omg_watcher/test/support/exit_processor/test_helper.ex [hunk 4]
```diff
@@ -42,8 +42,15 @@ defmodule OMG.Watcher.ExitProcessor.TestHelper do
     eth_height = Keyword.get(opts, :eth_height, 2)
     exit_id = Keyword.get(opts, :exit_id, @exit_id)
     call_data = %{utxo_pos: enc_pos, output_tx: txbytes}
-
-    event = %{owner: owner, eth_height: eth_height, exit_id: exit_id, call_data: call_data}
+    root_chain_txhash = <<0::256>>
```
Comment (unnawut) [inline apps/omg_watcher/test/support/exit_processor/test_helper.ex:45]: >  test helper updated to include a transaction hash of `<<0::256>>`. Open question whether we'd like a more advanced logic there in this PR.

I think something like this works: https://github.com/omisego/elixir-omg/blob/master/apps/omg_watcher_info/test/support/factories/data_helper.ex#L28-L30. It's in omg_watcher_info namespace but it's straightforward enough to copy over into a `defp`.

Comment (okalouti) [reply apps/omg_watcher/test/support/exit_processor/test_helper.ex]: Thanks @unnawut. Created a `def` in this file that I also re-used for above. c35bec5

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex (hunk 3 above)
Comment (okalouti) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex]: Done 6287e2b.

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex (hunk 2 above)
Comment (okalouti) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex]: Done 6287e2b.

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex (hunk 1 above)
Comment (okalouti) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex]: Done 6287e2b.

**Comment by okalouti** [comment]
Also added dialyzer specs in `exit_info` with an addition of `module_t()` in `events.ex` to use for the spec of `make_event_data/3`

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex [hunk 5]
```diff
@@ -37,21 +47,26 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
...
           # this means the exit has been first seen active. If false, it won't be considered harmful
           is_active: boolean(),
-          eth_height: pos_integer()
+          eth_height: pos_integer(),
+          root_chain_txhash: Crypto.hash_t()
         }
 
-  def new(contract_status, %{eth_height: eth_height, call_data: %{output_tx: txbytes}, exit_id: exit_id} = event) do
+  @spec new(map(), map()) :: t()
+  def new(contract_status, event) do
```
[the remark this answers is not in the document]
Comment (InoMurko) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:55]: my comment (to use a `dot` access modifier was strictly for internal type structs `ExitInfo`

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex [hunk 6]
```diff
@@ -61,68 +76,49 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
...
-           exiting_txbytes: exiting_txbytes,
-           is_active: is_active,
-           eth_height: eth_height
-         }}
-      )
-      when is_integer(amount) and is_integer(eth_height) and
-             is_binary(currency) and is_binary(owner) and is_integer(exit_id) and is_binary(exiting_txbytes) and
-             is_boolean(is_active) do
+  @spec make_db_update({Utxo.Position.t(), t()}) :: Utxo.Position.db_t()
+  def make_db_update({position, event}) do
```
Comment (InoMurko) [inline apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:95]:
```suggestion
  def make_db_update({position, exit_info}) do
```

Comment (okalouti) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:95]: Both make sense to me but `exit_info` is clearer. Changed in other methods for consistency except using `exit_event` for `new/2`

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex [hunk 7]
```diff
@@ -61,68 +76,49 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
...
     do: Utxo.Position.decode!(utxo_pos_enc)
 
+  @spec do_new(map(), list(map())) :: t()
   defp do_new(contract_status, fields) do
     fields = Keyword.put_new(fields, :is_active, parse_contract_exit_status(contract_status))
     struct!(__MODULE__, fields)
   end
 
+  @spec make_event_data(Event.module_t(), Utxo.Position.t(), t()) :: struct()
   def make_event_data(type, position, %__MODULE__{} = exit_info) do
```
[the remark this answers is not in the document]
Comment (okalouti) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex]: 999b0d5

File: apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (okalouti) [reply apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex]: clear reverted back thanks

**Pull request merged**